### PR TITLE
[SC-299] Add wsteth and reth exchange rate oracles

### PR DESCRIPTION
**Pull request by hexonaut**
Oracles which use the exchange rate instead of market rate for pricing for both wstETH and rETH. I removed the integration tests that are no longer valid since we've started upgrading the markets and those tests are obsolete now as I've updated the block number to get the most recent WSTETH oracle.

### CODE REVIEW

File: src/interfaces/IPriceSource.sol [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity >=0.8.0;
+
+interface IPriceSource {
+    function latestAnswer() external view returns (int256);
+    function decimals() external view returns (uint8);
+}
```
Comment (barrutko) [inline src/interfaces/IPriceSource.sol]: Missing empty line at the end

File: test/SparkLendMainnetIntegration.t.sol [hunk 2]
```diff
@@ -158,32 +144,7 @@ contract SparkLendMainnetIntegrationTest is Test {
         _triggerUpdate(ETH);
 
         // slope1 has adjusted down a bit so the borrow rate is slightly lower at same utilization
-        assertEq(_getBorrowRate(ETH), 0.027800738894650680218981526e27);
-    }
-
-    function test_usdc_usdt_market_oracles() public {
```
Comment (barrutko) [inline test/SparkLendMainnetIntegration.t.sol:164]: Why was this test removed?

**Review by barrutko — COMMENTED** [review]
Generally looks good. Happy to approve after resolving two minor comments.

File: test/SparkLendMainnetIntegration.t.sol (hunk 2 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:164]: I had to bump the block, and we've already applied these upgrades to the market.

File: src/interfaces/IPriceSource.sol (hunk 1 above)
Comment (hexonaut) [reply src/interfaces/IPriceSource.sol]: fixed

File: test/SparkLendMainnetIntegration.t.sol (hunk 2 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:164]: I could re-add, but not sure if it provides much value.

File: src/RETHExchangeRateOracle.sol [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        // 8 decimals required as AaveOracle assumes this
+        require(IPriceSource(_ethSource).decimals() == 8, "RETHExchangeRateOracle/invalid-decimals");
+        
+        reth      = IRocketPoolStakedEth(_reth);
+        ethSource = IPriceSource(_ethSource);
+    }
+
+    function latestAnswer() external view returns (int256) {
+        int256 ethUsd = ethSource.latestAnswer();
+        int256 exchangeRate = int256(reth.getExchangeRate());
```
Comment (lucas-manuel) [inline src/RETHExchangeRateOracle.sol:33]:
```suggestion
        int256 ethUsd       = ethSource.latestAnswer();
        int256 exchangeRate = int256(reth.getExchangeRate());
```

File: test/RETHExchangeRateOracle.t.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    function test_latestAnswer_zeroExchangeRate() public {
+        reth.setExchangeRate(0);
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    function test_latestAnswer_negativeExchangeRate() public {
+        // RETH ER can't go negative, but it can have a silent overflow
+        assertLt(int256(uint256(type(int256).min)), 0);
```
Comment (lucas-manuel) [inline test/RETHExchangeRateOracle.t.sol]: Why not use -1 here?

File: test/RETHExchangeRateOracle.t.sol [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        reth.setExchangeRate(uint256(type(int256).min));
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    function test_latestAnswer() public {
+        assertEq(oracle.latestAnswer(), 2400e8);
```
Comment (lucas-manuel) [inline test/RETHExchangeRateOracle.t.sol:78]: Can we add a few more hardcoded examples to demo the exchange rate multiplication?

File: src/RETHExchangeRateOracle.sol [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ *  @title RETHExchangeRateOracle
+ *  @dev   Provides rETH / USD by multiplying the rETH exchange rate by ETH / USD.
+ *         This provides a "non-market" price. Any depeg event will be ignored.
+ */
+contract RETHExchangeRateOracle {
+
+    /**
+     * @notice RocketPool staked eth token contract.
+     */
```
Comment (lucas-manuel) [inline src/RETHExchangeRateOracle.sol]: Can do /// @notice for these

File: src/WSTETHExchangeRateOracle.sol [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        // 8 decimals required as AaveOracle assumes this
+        require(IPriceSource(_ethSource).decimals() == 8, "WSTETHExchangeRateOracle/invalid-decimals");
+        
+        steth     = ILidoStakedEth(_steth);
+        ethSource = IPriceSource(_ethSource);
+    }
+
+    function latestAnswer() external view returns (int256) {
+        int256 ethUsd = ethSource.latestAnswer();
+        int256 exchangeRate = int256(steth.getPooledEthByShares(1e18));
```
Comment (lucas-manuel) [inline src/WSTETHExchangeRateOracle.sol:33]:
```suggestion
        int256 ethUsd       = ethSource.latestAnswer();
        int256 exchangeRate = int256(steth.getPooledEthByShares(1e18));
```

File: test/WSTETHExchangeRateOracle.t.sol [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        ethSource.setLatestAnswer(0);
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    function test_latestAnswer_negativeEthUsd() public {
+        ethSource.setLatestAnswer(-1);
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    function test_latestAnswer_zeroExchangeRate() public {
```
Comment (lucas-manuel) [inline test/WSTETHExchangeRateOracle.t.sol:64]: Why no negative exchange rate test here?

File: test/WSTETHExchangeRateOracle.t.sol [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        steth.setExchangeRate(0);
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    function test_latestAnswer() public {
+        assertEq(oracle.latestAnswer(), 2400e8);
```
Comment (lucas-manuel) [inline test/WSTETHExchangeRateOracle.t.sol:74]: Couple more hardcoded examples

File: test/SparkLendMainnetIntegration.t.sol (hunk 2 above)
Comment (lucas-manuel) [reply test/SparkLendMainnetIntegration.t.sol:164]: I think its a good idea to keep these tests and set the block within the test

File: test/SparkLendMainnetIntegration.t.sol [hunk 10]
```diff
@@ -32,43 +32,29 @@ contract SparkLendMainnetIntegrationTest is Test {
...
     IAaveOracle            aaveOracle            = IAaveOracle(AAVE_ORACLE);
     IPoolAddressesProvider poolAddressesProvider = IPoolAddressesProvider(POOL_ADDRESSES_PROVIDER);
     IPool                  pool                  = IPool(POOL);
     IPoolConfigurator      configurator          = IPoolConfigurator(POOL_CONFIGURATOR);
 
     function setUp() public {
-        vm.createSelectFork(getChain("mainnet").rpcUrl, 18784436);  // Dec 14, 2023
-    }
-
-    function test_dai_market_oracle() public {
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol:54]: Same here an keep the test and set the block within it

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/RETHExchangeRateOracle.sol (hunk 3 above)
Comment (hexonaut) [reply src/RETHExchangeRateOracle.sol:33]: fixed

File: test/RETHExchangeRateOracle.t.sol (hunk 4 above)
Comment (hexonaut) [reply test/RETHExchangeRateOracle.t.sol]: Ya good point that's simpler. Changed.

File: test/RETHExchangeRateOracle.t.sol (hunk 5 above)
Comment (hexonaut) [reply test/RETHExchangeRateOracle.t.sol:78]: done

File: src/RETHExchangeRateOracle.sol (hunk 6 above)
Comment (hexonaut) [reply src/RETHExchangeRateOracle.sol]: fixed

File: src/WSTETHExchangeRateOracle.sol (hunk 7 above)
Comment (hexonaut) [reply src/WSTETHExchangeRateOracle.sol:33]: fixed

File: test/WSTETHExchangeRateOracle.t.sol (hunk 8 above)
Comment (hexonaut) [reply test/WSTETHExchangeRateOracle.t.sol:64]: Unlike RETH you can't set the exchange rate to negative due to the multiplication in stETH. I made a note about that.

File: test/WSTETHExchangeRateOracle.t.sol (hunk 9 above)
Comment (hexonaut) [reply test/WSTETHExchangeRateOracle.t.sol:74]: done

File: test/SparkLendMainnetIntegration.t.sol (hunk 2 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:164]: ok re-added

File: test/SparkLendMainnetIntegration.t.sol (hunk 10 above)
Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol:54]: readded

File: test/WSTETHExchangeRateOracle.t.sol [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        ethSource.setLatestAnswer(-1);
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    function test_latestAnswer_zeroExchangeRate() public {
+        steth.setExchangeRate(0);
+        assertEq(oracle.latestAnswer(), 0);
+    }
+
+    // Note: exchange rate cannot overflow to negative because it multiplies by
```
Comment (lucas-manuel) [inline test/WSTETHExchangeRateOracle.t.sol]: Exchange

Comment (hexonaut) [reply test/WSTETHExchangeRateOracle.t.sol]: fixed

File: test/SparkLendMainnetIntegration.t.sol [hunk 12]
```diff
@@ -257,9 +271,9 @@ contract SparkLendMainnetIntegrationTest is Test {
...
         _triggerUpdate(USDT);
 
+        assertEq(_getBorrowRate(USDC), currentRateUSDC + 0.002000000000000000000000000e27, "after2: USDC mismatch");
         // Almost +0.2%, but utilization is slightly below the kink point
-        assertEq(_getBorrowRate(USDC), currentRateUSDC + 0.001891955013226013824859422e27, "after2: USDC mismatch");
-        assertEq(_getBorrowRate(USDT), currentRateUSDT + 0.001899234005225860815063204e27, "after2: USDT mismatch");
+        assertEq(_getBorrowRate(USDT), currentRateUSDT + 0.001602551612415803144238481e27, "after2: USDT mismatch");
     }
 
     function _getBorrowRate(address asset) internal view returns (uint256) {
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: We should move the new tests above these and add a section label to indicate these are helpers

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/SparkLendMainnetIntegration.t.sol [hunk 13]
```diff
@@ -343,4 +322,27 @@ contract SparkLendMainnetIntegrationTest is Test {
         assertEq(aaveOracle.getSourceOfAsset(WSTETH), address(oracle));
     }
 
+    // ---- Helper Functions ----
```
Comment (lucas-manuel) [inline test/SparkLendMainnetIntegration.t.sol]: Should maintain consistency with rest of codebase with

```
    /**********************************************************************************************/
    /*** Declarations and Constructor                                                           ***/
    /**********************************************************************************************/
```

styling

Comment (hexonaut) [reply test/SparkLendMainnetIntegration.t.sol]: done

**Review by lucas-manuel — APPROVED** [review]

**Review by barrutko — APPROVED** [review]

**Pull request merged**